org.eclipse.xtext.xbase/src/org/eclipse/xtext/xbase/compiler/XbaseCompiler.java

```diff
@@ -644,11 +644,40 @@ protected JvmTypeReference resolveSynonymType(JvmSynonymTypeReference reference,
 	/**
 	 * @since 2.18
 	 */
-	protected void appendFinallyWithResources(List<XVariableDeclaration> resources, ITreeAppendable b) {
-		for (int i = resources.size() - 1; i >= 0; i--) {
-			XVariableDeclaration res = resources.get(i);
-			String resName = getVarName(res, b);
-			b.newLine().append("if (" + resName + " != null) ").append(resName + ".close();");
+	protected void appendFinallyWithResources(XTryCatchFinallyExpression expr, List<XVariableDeclaration> resources,
+			ITreeAppendable b) {
+		if (expr != null && !resources.isEmpty()) {
+			// Invoking the close method might throw. Hence, we collect those
+			// Throwables and propagate them later on.
+			String throwablesStore = b.declareSyntheticVariable(Tuples.pair(expr, "_catchedThrowables"), "_ts");
+			b.newLine().append("java.util.List<Throwable> ");
+			b.append(throwablesStore);
+			b.append(" = new java.util.ArrayList<Throwable>();");
+			for (int i = resources.size() - 1; i >= 0; i--) {
+				XVariableDeclaration res = resources.get(i);
```

Comment:
We could use openPseudoScope on the appendable to avoid unnecessary counters on the catch-vars in the generated code

Reply:
Done.